internal/backend/remote-state/s3/client.go

```diff
@@ -295,6 +304,72 @@ func (c *RemoteClient) Lock(info *statemgr.LockInfo) (string, error) {
 
 	log.Info("Locking remote state")
 
+	var fileLockID, dynamoDBLockID string
+	if c.useLockFile {
+		fileLockID, err = c.lockWithFile(ctx, info, log)
+		if err != nil {
+			return "", err
+		}
+	}
+
+	if c.ddbTable != "" {
+		dynamoDBLockID, err = c.lockWithDynamoDB(ctx, info, log)
+		if err != nil {
+			// Release the file lock if attempting to acquire the DynamoDB lock fails.
+			if c.useLockFile {
+				if unlockErr := c.unlockWithFile(ctx, fileLockID, &statemgr.LockError{}, log); unlockErr != nil {
+					err = fmt.Errorf("error when attempting to clean up the file lock after failing to acquire DynamoDB lock: %v; original DynamoDB lock error: %w", unlockErr, err)
+				}
+			}
+			return "", err
+		}
+	}
+
+	// Combine file and DynamoDB lock IDs using a delimiter to support using both locking mechanisms at the same time.
+	return joinLockIDs(fileLockID, dynamoDBLockID), err
+}
+
+// lockWithFile attempts to acquire a lock on the remote state by uploading a lock file to Amazon S3.
+//
+// This method is used when the S3 native locking mechanism is in use. It uploads a lock file (JSON)
+// to an S3 bucket to establish a lock on the state file. If the lock file does not already
+// exist, the operation will succeed, acquiring the lock. If the lock file already exists, the operation
+// will fail due to a conditional write, indicating that the lock is already held by another Terraform client.
+func (c *RemoteClient) lockWithFile(ctx context.Context, info *statemgr.LockInfo, log hclog.Logger) (string, error) {
+	lockFileJson, err := json.Marshal(info)
+	if err != nil {
+		return "", err
+	}
+
+	input := &s3.PutObjectInput{
+		ContentType: aws.String("application/json"),
+		Body:        bytes.NewReader(lockFileJson),
+		Bucket:      aws.String(c.bucketName),
+		Key:         aws.String(c.lockFilePath),
+		IfNoneMatch: aws.String("*"),
+	}
+
+	log.Debug("Uploading lock file")
+
+	uploader := manager.NewUploader(c.s3Client, func(u *manager.Uploader) {})
```

Comment:
The lock file will always be small, so we can use a direct call to `s3.PutObject` instead of going though the S3 Upload Manager

Reply:
fixed in [e246ad3](https://github.com/hashicorp/terraform/pull/35661/commits/e246ad32f9711d4433220426bba9ff4b2da5fe92)